### PR TITLE
Abstract search functions

### PR DESCRIPTION
**Pull request by AVakiliT**
The SearchListView now calls the Anserini for relevant docs.
I didn't implement a get_content function since reading documents from the disk with nginx is working. But we can add it if you think it's necessary.
I'm also not sure how to handle having an interface for search functions, since django doesn't have any kind of dependency injection. Should we have some kind of service provider class that loads class names from the config file?
I also changed the json FastAPI was returning so it would match what the view expected.

### CODE REVIEW

**Comment by KamyarGhajar** [comment]
Cheers mate. So, How we should activate the anserini as the search engine now? And where should we put the anserini? Shouldn't it be another docker container or so?

**Comment by AVakiliT** [comment]
Mustafas already done all of the heavy lifting for that 😂. Anserini is in a separate docker container that reads an index of the disk the API is in search/app/main.py.

**Comment by KamyarGhajar** [comment]
Thanks to @ammsa too. I have checked the changes, they seem to be fine. What are the exact steps to test your changes?

**Comment by ammsa** [comment]
@AVakiliT Great thanks! I'll take a look and I might ask for some changes.

File: web/Web/web/interfaces/SearchEngine/anserini.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+from web.interfaces.SearchEngine.base import Search
+from collections import OrderedDict
+
+
+class Anserini(Search):
+
+    @staticmethod
+    def search(query: str, size: int):
+        response = requests.get(
+            f"http://{SEARCH_SERVER_IP}:{SEARCH_SERVER_PORT}/search",
```
Comment (ammsa) [inline web/Web/web/interfaces/SearchEngine/anserini.py:13]: replace these two variables with `self.host` and `self.port`

File: web/Web/web/interfaces/SearchEngine/anserini.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            params={
+                "query": query,
+                "size": size
+            }
+        )
+        response.raise_for_status()
+        results = response.json()
```
Comment (ammsa) [inline web/Web/web/interfaces/SearchEngine/anserini.py:20]: We can just return `results` here after we change the return type.

File: web/Web/web/interfaces/SearchEngine/anserini.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        document_values = OrderedDict(results)
+        document_ids = document_values.keys()
+        return document_ids, document_values, response.elapsed.total_seconds()
+
+    @staticmethod
+    def get_content(docno: str) -> str:
+        raise NotImplementedError
```
Comment (ammsa) [inline web/Web/web/interfaces/SearchEngine/anserini.py:27]: `get_content` shouldn't be hard to implement. For bigger collections that are usually compressed, nginx would be hassle to get it fetching document content.

File: search/app/main.py [hunk 4]
```diff
@@ -21,19 +19,27 @@ def read_root():
...
                 "score": hits[i].score,
                 "title": searcher.doc(hits[i].docid).contents()[:15],
                 "snippet": searcher.doc(hits[i].docid).contents()[:50]
             }
-            for i in range(len(hits))
-        ]
+        for i in range(len(hits))
     }
+
+# @app.get("/content/{id}")
```
Comment (ammsa) [inline search/app/main.py:37]: change `id` to `docno`

File: search/app/main.py [hunk 5]
```diff
@@ -21,19 +19,27 @@ def read_root():
...
+# def get_content(id: str):
+#     doc = searcher.doc(id)
+#     if doc is None:
+#         raise HTTPException(status_code=404, detail="Doc not found")
+#     return {
+#         "docno": id,
+#         "content": doc.content
```
Comment (ammsa) [inline search/app/main.py:44]: `doc.content()`

**Review by ammsa — CHANGES_REQUESTED** [review]
Hi @AVakiliT 

Thanks for working on this! I added some minor changes. 

`CALEngine/bmi_fcgi` file shouldn't be part of the commit.

**Comment by ammsa** [comment]
> Thanks to @ammsa too. I have checked the changes, they seem to be fine. What are the exact steps to test your changes?

@KamyarGhajar The updates are not yet visible in the browser but if you can fetch the changes and bring up the search container for now. `docker-compose logs -f --tail 100 django` to see if there's any errors while trying to search for document

File: web/Web/web/search/views.py [hunk 6]
```diff
@@ -5,10 +5,10 @@
...
 from django.template import loader
 from django.views import generic
-import httplib2
+import requests
 
 from web.interfaces.DocumentSnippetEngine import functions as DocEngine
-from web.interfaces.SearchEngine import functions as SearchEngine
+from web.interfaces.SearchEngine.anserini import Anserini as SearchEngine
 from web.search import helpers
 from web.core.mixin import RetrievalMethodPermissionMixin
```
[the remark this answers is not in the document]
Comment (AVakiliT) [reply web/Web/web/search/views.py:16]: I ended up putting a setting in the config file which is a string pointing to the anserini search implementation. The view imports this string so we can easily swap it out for another class if needed.

File: web/Web/web/interfaces/SearchEngine/base.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+class Search(ABC):
+
+    def __init__(self, host, port):
+        self.host = host
+        self.port = port
+        super().__init__()
+
+    @staticmethod
+    @abstractmethod
+    def search(query: str, size: int) -> (list, dict, float):
```
[the remark this answers is not in the document]
Comment (AVakiliT) [reply web/Web/web/interfaces/SearchEngine/base.py]: The api now returns something exactly like this.

**Comment by AVakiliT** [comment]
I made the requested changes.
I should not that the search api and interface have content and raw functions. But refactoring the codebase to use them should probably be a separate branch.

**Pull request merged**